File: 0501~1000/0894. All Possible Full Binary Trees/main.cpp
/** 
 * Source: https://leetcode.com/problems/all-possible-full-binary-trees/
 * Date: 2021/7/22
 * Skill: Recursion
 * Runtime: 112 ms, faster than 34.90% of C++ online submissions
 * Memory Usage: 38.3 MB, less than 25.29% of C++ online submissions 
 * Time complexity:
 * Space complexity: 
 * Constraints: 
 * 
 */

#include<iostream>
#include<vector>
#include<string>
using namespace std;

//* Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int n) {
        vector<TreeNode*> ans;
        if (n == 1) 
            ans.push_back(new TreeNode(0));
        else {
            if (n%2 == 0) return {};
            for (int i = 1; i <= n-2; i+=2) {
                if (i%2 == 1 && (n-i-1)%2 == 1) {
                    for (auto j:allPossibleFBT(i)) {
                        for (auto k:allPossibleFBT(n-i-1)) {
                            TreeNode* root = new TreeNode(0);
                            root->left = j;
                            root->right = k;
                            ans.push_back(root);
                        }
                    }
                }
            }
        }
        return ans;
    }
};

static const auto io_sync_off = []() {
    //* turn off sync
    std::ios::sync_with_stdio(false);
    //* untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0501~1000/0894. All Possible Full Binary Trees/main.py
"""
Source: https://rb.gy/17gsq
Date: 2023/7/23
Skill:
Ref:
Runtime: 206 ms, faster than 36.49%
Memory Usage: 26.05 MB, less than 20.07%
Time complexity:
Space complexity:
Constraints:
    1 <= n <= 20
"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        if n % 2 == 0:
            return None
        elif n == 1:
            return [TreeNode()]
        else:
            res = []
            for i in range(1, n-1, 2):
                left = self.allPossibleFBT(i)
                right = self.allPossibleFBT(n-1-i)
                for left_node in left:
                    for right_node in right:
                        root = TreeNode(left=left_node, right=right_node)
                        res.append(root)
            return res


if __name__ == "__main__":
    s = Solution()
